tests/run-macros/typeclass-derivation2d/Lib_1.scala

```diff
@@ -0,0 +1,343 @@
+              unpickleCase[`alt` & T, m.ElemTypes](buf, m)
+          }
+        else unpickleCases[T, alts1](n + 1)(buf, ord)
+      case _: Unit =>
+        throw new IndexOutOfBoundsException(s"unexpected ordinal number: $ord")
+    }
+
+  inline def derived[T](implicit ev: Mirror.Of[T]): Pickler[T] = new {
+    def pickle(buf: mutable.ListBuffer[Int], x: T): Unit =
+      inline ev match {
+        case m: Mirror.SumOf[T] =>
+          val ord = m.ordinal(x)
+          buf += ord
+          pickleCases[m.ElemTypes](0)(buf, x, ord)
+        case m: Mirror.ProductOf[T] =>
+          pickleElems[m.ElemTypes](0)(buf, x)
+      }
+    def unpickle(buf: mutable.ListBuffer[Int]): T =
+      inline ev match {
+        case m: Mirror.SumOf[T] =>
+          val ord = nextInt(buf)
+          unpickleCases[T, m.ElemTypes](0)(buf, ord)
+        case m: Mirror.ProductOf[T] =>
+          unpickleCase[T, m.ElemTypes](buf, m)
+      }
+  }
+
+  implicit object IntPickler extends Pickler[Int] {
+    def pickle(buf: mutable.ListBuffer[Int], x: Int): Unit = buf += x
+    def unpickle(buf: mutable.ListBuffer[Int]): Int = nextInt(buf)
+  }
+}
+
+// ----------- A third typeclass, making use of labels --------------------------
+
+trait Show[T] {
+  def show(x: T): String
+}
+object Show {
+  import scala.compiletime.{erasedValue, constValue, summonFrom}
+
+  inline def tryShow[T](x: T): String =
+    ${ tryShowExpr('x) }
+
+  private def tryShowExpr[T: Type](x: Expr[T])(given qctx: QuoteContext): Expr[String] = {
+    searchImplicitExpr[Show[T]] match {
+      case Some(s) => '{ $s.show($x) }
+      case None =>
+        qctx.error(s"Could not find delegate for Show[${summon[Type[T]].show}]")
+        '{ "" }
+    }
+  }
+
+  inline def showElems[Elems <: Tuple, Labels <: Tuple](n: Int)(x: Any): List[String] =
+    ${ unpickleElemsExpr[Elems, Labels]('n)('x) }
+
+  private def unpickleElemsExpr[Elems <: Tuple : Type, Labels <: Tuple : Type](n: Expr[Int])(x: Expr[Any])(given QuoteContext): Expr[List[String]] =
+    '{ ??? : Elems } match {
+      case '{ type $elems1 <: Tuple; $_ : ($elem *: `$elems1`) } =>
+        '{ ??? : Labels } match {
+          case '{ type $labels1 <: Tuple; $_ : ($label *: `$labels1`) } =>
+            val Const.Expr(label0) = label
+            '{
+              val formal = $label0
+              val actual = ${ tryShowExpr('{productElement[$elem]($x, $n)}) }
+              s"$formal = $actual" :: ${ unpickleElemsExpr('{$n + 1})(x)(given elems1, labels1, summon) }
+            }
+        }
+      case '{ $_ : Unit } =>
+        '{ Nil }
+    }
+
+  inline def showCase(x: Any, m: Mirror.ProductOf[_]): String = {
+    val label = constValue[m.CaseLabel]
+    inline m match {
+      case m: Mirror.Singleton => label
+      case _ => showElems[m.ElemTypes, m.ElemLabels](0)(x).mkString(s"$label(", ", ", ")")
+    }
+  }
+
+  inline def showCases[Alts <: Tuple](n: Int)(x: Any, ord: Int): String =
+    inline erasedValue[Alts] match {
```

Comment:
What is `erasedValue` necessary?

Reply:
That comes from the original implementation. I have not ported it yet. It is a way to encode a match on a type.